### PR TITLE
Add support for enhanced sound.

### PR DESCRIPTION
**Pull request by jg18**
Add support for enhanced in-mission sound, with 128 audio channels and a
new system for specifying per-sound priorities and concurrent playback
limits using modular extensions of soudns.tbl.

### CODE REVIEW

File: code/sound/ds.cpp [hunk 1]
```diff
@@ -1444,6 +1477,169 @@ int ds_get_free_channel(float new_volume, int snd_id, int priority)
...
+		}
+	}
+
+	if ( (first_free_channel >= 0) && (Channels[first_free_channel].source_id == 0) ) {
+		OpenAL_ErrorCheck( alGenSources(1, &Channels[first_free_channel].source_id), return -1 );
+	}
+	return first_free_channel;
+}
+
+/**
```
Comment (Echelon9) [inline code/sound/ds.cpp]: What function does this comment block relate to? Can it be deleted?

File: code/cmdline/cmdline.cpp [hunk 2]
```diff
@@ -167,6 +167,7 @@ Flag exe_params[] =
 	{ "-snd_preload",		"Preload mission game sounds",				true,	EASY_MEM_ALL_ON,	EASY_DEFAULT_MEM,	"Audio",		"http://www.hard-light.net/wiki/index.php/Command-Line_Reference#-snd_preload", },
 	{ "-nosound",			"Disable all sound",						false,	0,					EASY_DEFAULT,		"Audio",		"http://www.hard-light.net/wiki/index.php/Command-Line_Reference#-nosound", },
 	{ "-nomusic",			"Disable music",							false,	0,					EASY_DEFAULT,		"Audio",		"http://www.hard-light.net/wiki/index.php/Command-Line_Reference#-nomusic", },
+	{ "-enhanced_sound", "Enable enhanced sound", false, 0, EASY_DEFAULT, "Audio", "http://www.hard-light.net/wiki/index.php/Command-Line_Reference#-enhanced_sound", },
```
Comment (Echelon9) [inline code/cmdline/cmdline.cpp]: Whitespace should be amended, so that the text aligns with the lines above.

File: code/gamesnd/gamesnd.cpp [hunk 3]
```diff
@@ -14,12 +14,250 @@
...
+	EnhancedSoundData(  SND_ENHANCED_PRIORITY_MUST_PLAY,  1), // SND_ENGINE                     = 4,  //!< engine sound (as heard in cockpit)
+	EnhancedSoundData(  SND_ENHANCED_PRIORITY_MUST_PLAY,  1), // SND_CARGO_REVEAL               = 5,  //!< cargo revealed
+	EnhancedSoundData(  SND_ENHANCED_PRIORITY_MUST_PLAY,  3), // SND_DEATH_ROLL                 = 6,  //!< ship death roll
+	EnhancedSoundData(  SND_ENHANCED_PRIORITY_MUST_PLAY,  2), // SND_SHIP_EXPLODE_1             = 7,  //!< ship explosion 1
+	EnhancedSoundData(  SND_ENHANCED_PRIORITY_MUST_PLAY,  1), // SND_TARGET_ACQUIRE             = 8,  //!< target acquried
+	EnhancedSoundData(  SND_ENHANCED_PRIORITY_MUST_PLAY,  1), // SND_ENERGY_ADJUST              = 9,  //!< energy level change success
+	EnhancedSoundData(  SND_ENHANCED_PRIORITY_MUST_PLAY,  1), // SND_ENERGY_ADJUST_FAIL         = 10, //!< energy level change fail
+	EnhancedSoundData(  SND_ENHANCED_PRIORITY_MUST_PLAY,  1), // SND_ENERGY_TRANS               = 11, //!< energy transfer success
+	EnhancedSoundData(  SND_ENHANCED_PRIORITY_MUST_PLAY,  1),  // SND_ENERGY_TRANS_FAIL         = 12, //!< energy transfer fail
```
Comment (Echelon9) [inline code/gamesnd/gamesnd.cpp]: Whitespace should be amended, such that ``// SND_ENERGY_TRANS_FAIL`` aligns with the start of comments on the lines above.

File: code/sound/ds.cpp [hunk 4]
```diff
@@ -1024,6 +1028,25 @@ void ds_init_buffers()
...
 
 /**
+* Check if the player is using OpenAL Soft,
+* which is required to use enhanced sound.
+* Returns true on success, false otherwise.
+*/
+bool ds_check_for_openal_soft()
+{	
+	const ALchar * renderer = alGetString(AL_RENDERER);
+	Assertion(renderer!= NULL, "ds_check_for_openal_soft: renderer is null!");
```
Comment (asarium) [inline code/sound/ds.cpp]: Assertions should be used for programming errors but if the OpenAL implementation returns something invalid here it's a recoverable data error so I think it's better to just print a message to the log.

**Comment by asarium** [comment]
Currently the new code is only used if a commandline flag is set but that's a bit counterintuitive. I would propose changing it so the enhanced code is always enabled and if OpenALSoft is not detected it would be disabled and a message is written to the debug log. A `-no-enhanced-sound` option can be added to always disable the new code.

File: code/sound/ds.cpp (hunk 1 above)
Comment (jg18) [inline code/sound/ds.cpp]: I'm not sure what you're referring to. The fragment looks to be the end of get_free_channel_enhanced(), which attempts to trash the "least important" sound, generates the OpenAL source if needed, and returns the channel found (-1 on failure).

File: code/cmdline/cmdline.cpp (hunk 2 above)
Comment (jg18) [inline code/cmdline/cmdline.cpp]: Odd, the text looks aligned on this page but not in VS2013. As such, I have no idea how to align it in VS2013. Maybe the number of spaces per tab is wrong? Or there's some other setting in VS2013 that's wrong that's preventing me from seeing the text aligned?

Comment (chief1983) [inline code/cmdline/cmdline.cpp]: Instead of just going by looks, might need to verify whether the other lines are using spaces vs tabs.

Comment (jg18) [inline code/cmdline/cmdline.cpp]: Problem tracked down. I was using a non-monospace font (Arial Black) so the text wasn't aligned in VS2013. Switched to Consolas and now I'm able to fix it.

**Comment by The-E** [comment]
That would indeed be preferable, as it follows the precedent set by the various graphics flags.

**Comment by chief1983** [comment]
Actually the graphics flags are currently a grab bag of some disable, some enable, whatever.  I'm not sure I'm really a fan of that because it makes just reading the checklist difficult, but I do agree that since that's the case, for this flag, enabling it when possible by default would make the most sense.

File: code/cmdline/cmdline.cpp (hunk 2 above)
Comment (jg18) [inline code/cmdline/cmdline.cpp]: Fixed in commit faa0c5e.

File: code/gamesnd/gamesnd.cpp (hunk 3 above)
Comment (jg18) [inline code/gamesnd/gamesnd.cpp]: Fixed in commit faa0c5e.

File: code/sound/ds.cpp (hunk 4 above)
Comment (jg18) [inline code/sound/ds.cpp]: Fixed in commit faa0c5e.

File: code/sound/ds.cpp (hunk 1 above)
Comment (Echelon9) [inline code/sound/ds.cpp]: Apologies, I should have commented on the end, not start, of the comment block so these inline boxes were clearer.
I'm referring to the comment block that starts with "Generic function for getting a free channel".

It appears a pre-Doxygen layout, yet there is no function below it before the next Doxygen comment block which relates to ```ds_get_free_channel()```. If Doxygen-ified, can the older function description be removed as redundant?

**Comment by The-E** [comment]
This is good to go in then.

**Pull request merged**